client/my-sites/stats/activity-log-tasklist/index.jsx

```diff
@@ -0,0 +1,413 @@
+		} );
+	};
+
+	/**
+	 * Goes to general plugin management screen.
+	 *
+	 * @returns {object} Action to redirect to plugins management.
+	 */
+	goManagePlugins = () => this.props.goManagePlugins( this.props.siteSlug );
+
+	/**
+	 * Goes to single plugin management screen.
+	 *
+	 * @param {string} pluginSlug Plugin slug, like "hello-dolly".
+	 *
+	 * @returns {object} Action to redirect to plugin management.
+	 */
+	goToPlugin = pluginSlug => this.props.goToPlugin( pluginSlug, this.props.siteSlug );
+
+	/**
+	 * Checks if the plugin update queue has more items and none is currently updating.
+	 * If so, updates the next plugin.
+	 */
+	continueQueue = () => {
+		if (
+			0 < this.state.queuedPlugins.length &&
+			! isPluginUpdating( Object.values( this.props.pluginsWithUpdate ) )
+		) {
+			this.updatePlugin( this.state.queuedPlugins[ 0 ] );
+		}
+	};
+
+	/**
+	 * Add a plugin to the update queue.
+	 *
+	 * @param {object} plugin Plugin to enqueue.
+	 *
+	 * @returns {undefined}
+	 */
+	enqueuePlugin = plugin =>
+		this.setState(
+			{
+				queuedPlugins: union( this.state.queuedPlugins, [ plugin ] ),
+			},
+			this.continueQueue
+		);
+
+	/**
+	 * Remove a plugin from the update queue.
+	 *
+	 * @returns {undefined}
+	 */
+	dequeuePlugin = () =>
+		this.setState(
+			{
+				queuedPlugins: this.state.queuedPlugins.slice( 1 ),
+			},
+			this.continueQueue
+		);
+
+	/**
+	 * Add all plugins with pending updates to the queue and process it.
+	 *
+	 * @returns {undefined}
+	 */
+	updateAll = () =>
+		this.setState(
+			{
+				queuedPlugins: union(
+					this.state.queuedPlugins,
+					Object.values( this.props.pluginsWithUpdate )
+				),
+			},
+			this.continueQueue
+		);
+
+	/**
+	 * Starts the update process for a specified plugin. Displays an informational notice.
+	 *
+	 * @param {object} plugin Plugin information that includes
+	 * {
+	 * 		{string} id   Plugin id, like "hello-dolly/hello".
+	 * 		{string} slug Plugin slug, like "hello-dolly".
+	 * 		{string} name Plugin name, like "Hello Dolly".
+	 * }
+	 * @param {string} location Send 'from_task' when this is called a row in the task list,
+	 *                          'from_notice' when it's called from error notice.
+	 */
+	updatePlugin = ( plugin, location = 'from_task' ) => {
+		const { showInfoNotice, siteName, updateSinglePlugin } = this.props;
+
+		updateSinglePlugin( plugin, location );
+
+		showInfoNotice(
+			PluginNotices.inProgressMessage( 'UPDATE_PLUGIN', '1 site 1 plugin', {
+				plugin: plugin.name,
+				site: siteName,
+			} ),
+			{
+				id: plugin.slug,
+				showDismiss: false,
+			}
+		);
+	};
+
+	componentDidMount() {
+		const path = `/stats/activity/${ this.props.siteSlug }`;
+		page.exit( path, ( context, next ) => {
+			if (
+				! this.state.queuedPlugins.length ||
+				window.confirm( this.props.translate( 'Navigating away will cancel remaining updates' ) )
+			) {
+				return next();
+			}
+			page.replace( path, null, false, false );
```

Comment:
The `setTimeout( () => page.replace( ... ), 0 )` wrapper is most likely needed. All other Calypso implementations of the "navigation blocker" use it, and it's used also in internal implementation of `page.redirect`

Reply:
Ok I've restored it. Yes, the place I referenced this from in Calypso, had that setTimeout to 0. I tried with and without it, and experienced no difference, but I'm happy to leave it since it's in the other implementations.